src/protocol/mod.rs

```diff
@@ -348,8 +395,28 @@ impl WebSocketContext {
         }
 
         let frame = match message {
-            Message::Text(data) => Frame::message(data.into(), OpCode::Data(OpData::Text), true),
-            Message::Binary(data) => Frame::message(data, OpCode::Data(OpData::Binary), true),
+            Message::Text(data) => {
+                if let Some(pmce) = self.pmce.as_mut() {
+                    Frame::compressed_message(
+                        pmce.compress(data.as_bytes())?,
+                        OpCode::Data(OpData::Text),
+                        true,
+                    )
+                } else {
+                    Frame::message(data.into(), OpCode::Data(OpData::Text), true)
+                }
+            }
+            Message::Binary(data) => {
+                if let Some(pmce) = self.pmce.as_mut() {
+                    Frame::compressed_message(
+                        pmce.compress(&data)?,
+                        OpCode::Data(OpData::Binary),
+                        true,
+                    )
+                } else {
+                    Frame::message(data, OpCode::Data(OpData::Binary), true)
+                }
+            }
```

Comment:
This looks ok now when we only have 2 choices, but I think ideally to make implementation more flexible, we probably want to have some function that we call from here, which then handles the implementation details (extensions, if any) inside its body. What do you think?

Reply:
Extracted to a separate method `fn prepare_data_frame(&mut self, data: Vec<u8>, opdata: OpData) -> Result<Frame>` that handles both text and binary.